File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    static int answer;
    
    public int solution(int[] numbers, int target) {
        answer = 0;
        dfs(numbers, 0, target, 0);
        return answer;
    }
    
    public void dfs(int[] numbers, int depth, int target, int sum) {
        if (depth == numbers.length) {
            if (target == sum) {
                answer++;
            }
            return;
        }
        dfs(numbers, depth + 1, target, sum + numbers[depth]);
        dfs(numbers, depth + 1, target, sum - numbers[depth]);
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# DFS
def solution(numbers, target):
    answer = 0
    
    def dfs(cur, _sum):
        nonlocal answer
        
        if len(numbers) == cur:
            if _sum == target:
                answer += 1
            return
        
        dfs(cur + 1, _sum + numbers[cur])
        dfs(cur + 1, _sum - numbers[cur])
        
    dfs(0, 0)
    
    return answer